---
title: "Stoma analysis"
output: html_notebook
---


```{r}
knitr::opts_chunk$set(echo=T)
knitr::opts_chunk$set(include=T)
```


```{r library-load, include=FALSE}
#load required packages
library(qiime2R)
library(vegan)
library(phyloseq)
library(pairwiseAdonis)
library(tidyverse)
library(metagMisc)
library(ggprism)
library(rstatix)
```


```{r helper-set-up, include=FALSE}
theme_set(theme_classic(base_size = 16) + 
                     theme(axis.text = element_text(face = "bold", size = 15),
                           axis.title = element_text(face = "bold")))
#Define my colors
mycolors <- c("seagreen4","violetred4")
```

```{r import}
biom <-  import_biom("table-with-taxonomy.biom", parseFunction = parse_taxonomy_default)
metadata <- import_qiime_sample_data("Metadata.txt")#Then define variables as factors:
tree <- read_tree_greengenes("tree.nwk")
physeq <- merge_phyloseq(biom, metadata, tree)
colnames(tax_table(physeq)) <- c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")
physeq #make sure you have the right number of taxa and samples!
#Figure out which sample names are mismatched and adjust them
idx <- which(!(sample_names(biom) %in% sample_names(physeq)))
sample_names(biom) [idx]
sample_names(biom) [idx] <- str_to_lower(sample_names(biom) [idx])
physeq <- merge_phyloseq(biom, metadata, tree)
physeq
all(sample_names(biom) %in% sample_names(physeq))
```


```{r filtering}
##Filtering for samples:
#Filter samples with low depth (<1,000 reads):
sample_sums(physeq)
which(sample_sums(physeq) < 1000)
#Change this from Haggai's work, as not 8-prox also does not have enough counts for this cut-off
psfiltered <- prune_samples(sample_sums(physeq) >= 1000, physeq)
sample_names(psfiltered)
psfiltered
```


```{r beta-score}
#Calculate beta diversity distances
dist <- distance(psfiltered, method = "unifrac", type = "samples")
uwu_dist_df <- dist2list(dist) %>% 
  dplyr::rename("SampleID" = col, "ToSample" = row) %>% 
  filter(!str_detect(SampleID,"Eff"),!str_detect(ToSample,"Eff")) %>% 
  filter(str_trunc(as.character(SampleID), 2, side = "right", ellipsis = "") == 
           str_trunc(as.character(ToSample), 2, side = "right", ellipsis = "")) %>% 
  mutate(samp_dist = str_c(SampleID, str_sub(ToSample, start = 3)))

#Plot distance histogram
ggplot(uwu_dist_df) +
  geom_histogram(aes(value), binwidth = 0.1)
```

```{r permanova}
perm_meta <- as_tibble(sample_data(psfiltered), rownames = NA) %>% 
  mutate(Batch = as_factor(Batch))
vegan::adonis2(dist ~ Patient + Type + Batch +Sex + mode.of.delivery + Antibiotic_last_1_month_yes.no, method = "uwunifrac",
               data = perm_meta, by = "terms")
perm_res <- vegan::adonis2(dist ~ Patient + Type + Batch +Sex + mode.of.delivery + Antibiotic_last_1_month_yes.no, method = "uwunifrac",
               data = perm_meta, by = "terms")
#write_csv(as_tibble(perm_res, rownames = "Coef"), here::here("pubs_figures/uwunifrac_permanova with clinical variables.csv"))
pairwise.adonis2(dist ~ Type + Batch + Patient, method = "uwunifrac",
               data = perm_meta, by = "terms")
vegan::adonis2(dist ~ Patient + Type + Batch +Sex + mode.of.delivery + Antibiotic_last_1_month_yes.no,
               data = perm_meta, by = "margin")
vegan::adonis2(dist ~ Patient , data = perm_meta)
vegan::adonis2(dist ~ Type , data = perm_meta)
vegan::adonis2(dist ~ Batch , data = perm_meta)

anova(vegan::betadisper(dist, perm_meta$Patient))
anova(vegan::betadisper(dist, perm_meta$Type))
anova(vegan::betadisper(dist, perm_meta$Batch))
```






Plot final graphs to correlate metadata and Unifrac for publication - Type of stoma
Antibiotics in the last month
Time with stoma
Age at stoma closure
Mode of delivery
Nutritional independence at 30 days
Complications


```{r pub-figs_spear, fig.width=6, fig.height=4}
final_meta <- read_tsv(here::here("final_Metadata.txt"))
all(uwu_dist_df$SampleID %in% final_meta$`#SampleID`)
uwu_dist_pub <- uwu_dist_df %>% 
  left_join(final_meta, by = c("SampleID" = "#SampleID"))
uwu_dist_spear_cor_pub <- uwu_dist_pub %>% 
  select(value, c("Age at stoma closure (days)", "Time with stoma (days)", "Age at primary surgery (days)", "Last_weight",
                                     "Stoma volume (ml/day)"))
uwu_dist_spear_cor_pub_test <- uwu_dist_spear_cor_pub %>% 
  cor_test(vars = "value", vars2 = c("Age at stoma closure (days)", "Time with stoma (days)", "Age at primary surgery (days)", "Last_weight",
                                     "Stoma volume (ml/day)"), method = "spearman") %>% 
  mutate(padj = p.adjust(p, method = "BH"))
```

```{r pub-figs-box, fig.width=6, fig.height=4, eval=FALSE}
#Compute and plot Mann-Witney U tests
uwu_dist_box_pub <- uwu_dist_pub %>% 
  select(value, Sex, `PN use` ,`Antibiotic in last month of FU`, `Mode of delivery`, 
         `Nutritional independence at day 30`, Complication, `Type of stoma`) %>% 
  rename_with(.fn = function(x) str_replace_all(x, " ", "_"))
box_cats_pubs <- set_names(colnames(uwu_dist_box_pub) [-1], nm= colnames(uwu_dist_box_pub) [-1]) 
uwu_dist_box_pub_test <- map(box_cats_pubs, ~wilcox_test(data =uwu_dist_box_pub, as.formula(str_c("value ~ ", .)))) %>% 
  map(add_xy_position)

#Check adjusted p-values
dist_box_pub_adjust <- uwu_dist_box_pub_test %>% 
  reduce(rbind) %>% 
  mutate(padj = p.adjust(p, method = "BH")) 
#Combine all p-values for Spearman and for wilcoxon and adjust them together
raw_ps <- c(uwu_dist_spear_cor_pub_test$p, dist_box_pub_adjust$p)
adj_ps <- p.adjust(raw_ps, method = "BH")
#Put the adjusted ps in their appropriate tables
uwu_dist_spear_cor_pub_test$padj <- adj_ps [1:nrow(uwu_dist_spear_cor_pub_test)]
#write_csv(uwu_dist_spear_cor_pub_test, file = here::here(str_c("pubs_figures/U_Unifrac Spearman again test results.csv")))
dist_box_pub_adjust$padj <- adj_ps [(nrow(uwu_dist_spear_cor_pub_test)+1):length(adj_ps)]
#write_csv(dist_box_pub_adjust, file = here::here(str_c("pubs_figures/U_Unifrac Wilcoxon again test results.csv")))
dist_box_pub_adjust <- dist_box_pub_adjust %>% 
  split(., seq(nrow(.)))
```

```{r pub-figs_spear_plot, fig.width=6, fig.height=4}
#Plot correlations results
spear_plots <- map(c("Age at stoma closure (days)", "Time with stoma (days)"),
                   ~ggplot(uwu_dist_spear_cor_pub, aes_(x = quote(value), y = as.name(.x))) +
                     geom_smooth(method = "lm",color = mycolors [1], lwd = 1.5, fill = "grey70") +
                     geom_point(size = 3) +
                     labs(x = "Unweighted Unifrac Distance", 
                          subtitle = str_c("Spearman's Rank Test: \n", "Spearman's rho = ",
                                           pull(filter(uwu_dist_spear_cor_pub_test, var2 == .x), cor),
                                           " ; p = ", 
                                           round(pull(filter(uwu_dist_spear_cor_pub_test, var2 == .x), padj),2))) +
                     coord_cartesian(ylim = c(0,NA)) +
                     theme_classic(base_size = 16) + 
                     theme(axis.text = element_text(face = "bold", size = 15),
                           axis.title = element_text(face = "bold")))
spear_plots 
#Save the plots as pngs
#walk2(spear_plots, c("Age at stoma closure (days)", "Time with stoma (days)"), 
 #     ~ggsave(here::here(str_c("pubs_figures/U_Unifrac Spearman against ", .y,".png")),.x, width = 6, height = 4, dpi = 1080))
#walk2(spear_plots, c("Age at stoma closure (days)", "Time with stoma (days)"), 
 #     ~ggsave(here::here(str_c("pubs_figures/U_Unifrac Spearman against ", .y,".svg")),.x, width = 6, height = 4))
```

Make the figures for the boxplots

```{r pub-figs-box-plot, fig.width=6, fig.height=4, eval=FALSE}
#Plot the results

no_violin_p_num_ps <- map2(box_cats_pubs, uwu_dist_box_pub_test, 
                           ~ ggplot(uwu_dist_box_pub, aes_(x=as.name(.x), y=quote(value))) +
                             geom_boxplot(aes_(color = as.name(.x)), width = 0.7, alpha = 0, lwd=1) + 
                             ggforce::geom_sina(aes_(color = as.name(.x)),maxwidth = 0.4, size = 3.5, alpha = 1) +
                             #geom_jitter(aes_(color = as.name(.x)),width = 0.2, size = 3.5, alpha = 1) +
                             scale_color_manual(values = mycolors [c(1,2)]) +
                             guides(color = 'none') +
                             labs(y="Unweighted UniFrac\nDistance", x = str_replace_all(.x,"_"," ")) +
                             coord_cartesian(ylim = c(0,0.65)))
box_p_vals_ps <- pmap(list(no_violin_p_num_ps,dist_box_pub_adjust, names(no_violin_p_num_ps)),
                      ~..1 + add_pvalue(..2,
                                        bracket.nudge.y = 0.01,
                                        tip.length = 0,
                                        bracket.size = 1,
                                        label.size = 4.5,
                                        label = "p = {round(padj,2)}"))
box_p_vals_ps
#Print the results out as png and svg
#pwalk(list(box_p_vals_ps, names(no_violin_p_num_ps)),
 #     ~ggsave(here::here(str_c("pubs_figures/box number p U_Unifrac MW against ", ..2,".png")),
  #            ..1, width = 6, height = 4, dpi = 1080))
#pwalk(list(box_p_vals_ps, names(no_violin_p_num_ps)),
 #     ~ggsave(here::here(str_c("pubs_figures/box number p U_Unifrac MW against ", ..2,".svg")),
  #            ..1, width = 6, height = 4))

```


Load in protein data and calculate Euclidean distances between samples
```{r import-log-prot}
mq <- read_tsv(here::here("proteomics/normalyzer/test_run/log2-normalized.txt"))
#Fix the sample names
mq_meta <- read_tsv(here::here("proteomics/norm_meta.tsv")) %>% 
  mutate(sample_id = str_c(Patient,group, sep = "-")) %>% 
  mutate(sample_id = str_replace(sample_id,"14-prox","14-Prox"),
         sample_id = str_replace(sample_id,"14-dist","14-Dist"))
#Confirm sample names are fixed
all(mq_meta$sample_id %in% sample_names(physeq))
mq_mat <- mq [,-1] %>% 
  . [,mq_meta$sample] %>% 
  rename_with(function(x) mq_meta$sample_id) %>% 
  mutate(across(everything(),~ as.numeric(str_replace_na(.,replacement = "0")))) %>% 
  as.matrix()
rownames(mq_mat) <- mq$Genes
#Calculate Euclidean distances between samples
prot_dist <- dist(t(mq_mat), method = "euclidean", upper = T)
prot_dist_df <- dist2list(prot_dist, tri = F)  %>% 
  dplyr::rename("SampleID" = col, "ToSample" = row) %>%  
  filter(str_trunc(as.character(SampleID), 2, side = "right", ellipsis = "") == 
           str_trunc(as.character(ToSample), 2, side = "right", ellipsis = "")) %>% 
  filter(str_detect(SampleID, "(P|p)rox"), value > 0) %>% 
  mutate(samp_dist = str_c(SampleID, str_sub(ToSample, start = 3)))
#Plot distance histogram
ggplot(prot_dist_df) +
  geom_histogram(aes(value), binwidth = 50)
```


Repeat correlations with metadata for protein distances
```{r prot-distance-scatter}
#Add metadata variables to the table and correlate
prot_meta <- as_tibble(metadata)  %>% 
  mutate(Batch = as_factor(Batch))
prot_meta <- prot_meta %>% 
  mutate(sample.id = str_replace(sample.id, "Prox","prox"),
         sample.id = str_replace(sample.id, "Dist","dist"))
prot_dist_meta <- prot_dist_df %>% 
  left_join(prot_meta, by = c("SampleID" = "sample.id")) 
#Prepare and plot linear correlations
prot_dist_meta_pear <- prot_dist_meta %>% 
  select(colnames(prot_dist_df),where(is.numeric))
cors_cats <- colnames(prot_dist_meta_pear) [(ncol(prot_dist_df)+1):ncol(prot_dist_meta_pear)] 
#pdf(here::here("Protein Euclidean - numeric metadata correlations.pdf"), width = 8, height = 6)
walk(cors_cats, 
     ~print(ggstatsplot::ggscatterstats(prot_dist_meta_pear, x=value, y={{.}}, type = "nonparametric", 
                                        xlab = "Protein Euclidean Distance")))
#dev.off()
```

```{r prot-distance-box, eval=FALSE}
#Prepare and plot boxplots (categorical metadata)
prot_dist_meta_box <- prot_dist_meta %>% 
  select(colnames(prot_dist_df),where(is.character),where(is.factor),
         -Patient,-Type, -contains("Date"))
box_cats <- colnames(prot_dist_meta_box) [(ncol(prot_dist_df)+1):ncol(prot_dist_meta_box)] 
#pdf(here::here("Protein Euclidean - categoric metadata correlations.pdf"), width = 8, height = 6)
walk(box_cats, 
     ~print(ggstatsplot::ggbetweenstats(prot_dist_meta_box, x = {{.}}, y=value, 
                                        ylab = "Protein Euclidean Distance", type = "nonparametric")))
#dev.off()
```



```{r plot-prot-microbe, fig.width=6, fig.height=5}
#Calculate beta diversity distances - test all
method_list <- c("jaccard","bray","unifrac", "wunifrac") %>% 
  set_names(c("jaccard","bray","unifrac", "wunifrac"))
dist_list <- method_list %>% 
  map(~distance(psfiltered, method = ., type = "samples"))
dist_df_list <- dist_list %>% 
  map(dist2list) %>% 
  map(dplyr::rename,"SampleID" = col, "ToSample" = row) %>% 
  map(filter,!str_detect(SampleID,"Eff"),!str_detect(ToSample,"Eff")) %>%
  map(filter,str_trunc(as.character(SampleID), 2, side = "right", ellipsis = "") == 
           str_trunc(as.character(ToSample), 2, side = "right", ellipsis = "")) %>% 
  map(mutate,samp_dist = str_c(SampleID, str_sub(ToSample, start = 3)))


#Join microbial and protein distances
joint_dist_df_list <- dist_df_list %>% 
  map(left_join, prot_dist_df, by = c("SampleID","ToSample"))
p_list <- joint_dist_df_list %>% 
  map(~psych::corr.test(x=.$value.x, y = .$value.y,use = "pairwise", method="spearman", adjust="BH", ci=FALSE))
coef_top_list <- map(p_list,
                      ~reshape2::melt(.x [c(1,4)])) %>% 
  map(pivot_wider, names_from = "L1", values_from = "value") %>% 
  map2(names(p_list), ~mutate(.x, metric = .y))
#Repeat the analysis after omitting patient 1 and 2
joint_dist_df_filt_list <- joint_dist_df_list %>% 
  map(filter, str_detect(samp_dist.x, "(01)|(02)", negate = T))
p_filt_list <- joint_dist_df_filt_list %>% 
  map(~psych::corr.test(x=.$value.x, y = .$value.y,use = "pairwise", method="spearman", adjust="BH", ci=FALSE))
coef_top_filt_list <- map(p_filt_list,
                      ~reshape2::melt(.x [c(1,4)])) %>% 
  map(pivot_wider, names_from = "L1", values_from = "value") %>% 
  map2(names(p_list), ~mutate(.x, metric = .y))
#Print the plots
#pdf(here::here("pubs_figures/Distance correlations between 16S and proteome.pdf"),width = 6, height = 5)
pwalk(list(c(joint_dist_df_list,joint_dist_df_filt_list),
           rep(names(joint_dist_df_list), times = 2),
           c(coef_top_list, coef_top_filt_list),
           rep(c("All samples","Outlier removed"),each = length(joint_dist_df_list))),
           ~print(ggplot(..1, aes(x = value.x, y = value.y)) +
                    geom_smooth(method = "lm", color = mycolors [7], lwd = 1.5, fill = "grey70") + 
                    geom_point() + 
                    labs(x = str_c(str_to_title(..2)," 16S distance"), y ="Protein Euclidean distance",
                         title = str_c("Protein-microbe distance correlations ", ..4), 
                         subtitle = str_c("Spearman rho =",round(as.numeric(pull(..3,"r")),2)," p=",
                                          format(as.numeric(pull(..3,"p")),digits = 2, scientific = T)))))
pwalk(list(c(joint_dist_df_list,joint_dist_df_filt_list),
           rep(names(joint_dist_df_list), times = 2),
           c(coef_top_list, coef_top_filt_list),
           rep(c("All samples","Outlier removed"),each = length(joint_dist_df_list))),
           ~print(ggplot(..1, aes(x = value.x, y = value.y)) +
                    geom_smooth(method = "lm") +
                     ggrepel::geom_text_repel(aes(label = str_trunc(samp_dist.x, 2, "right", ellipsis = ""))) +
                    labs(x = str_c(str_to_title(..2)," 16S distance"), y ="Protein Euclidean distance",
                         title = str_c("Protein-microbe distance correlations ", ..4), 
                         subtitle = str_c("Spearman rho =",round(as.numeric(pull(..3,"r")),2)," p=",
                                          format(as.numeric(pull(..3,"p")),digits = 2, scientific = T)))))
#dev.off()

padj_coef_unifrac <- coef_top_list %>% 
  reduce(rbind) %>% 
  mutate(padj = p.adjust(p, method = "BH")) %>% 
  filter(metric == "unifrac")

uwu_euclid_plot <- ggplot(joint_dist_df_list [["unifrac"]], aes(x = value.x, y = value.y)) +
  geom_smooth(method = "lm",color = mycolors [1], lwd = 3, fill = "grey70") +
  geom_point(size = 4) +
  labs(x = str_c("UniFrac"," 16S\nDistance"), y ="Protein Euclidean\nDistance",
       subtitle = str_c("Spearman rho = ",round(as.numeric(pull(padj_coef_unifrac,"r")),2),", p = ",
                        format(as.numeric(pull(padj_coef_unifrac,"padj")),digits = 2, scientific = T))) +
  theme_classic(base_size = 20) + 
  theme(axis.text = element_text(face = "bold", size = 18),
        axis.title = element_text(face = "bold"))
uwu_euclid_plot
#ggsave(here::here("pubs_figures/Unifrac-Euclid spear_cor.png"), uwu_euclid_plot, width = 6, height = 5, dpi = 1080)
#ggsave(here::here("pubs_figures/Unifrac-Euclid spear_cor.svg"), uwu_euclid_plot, width = 6, height = 5)
```



